src/compiler/checker.ts

```diff
@@ -2585,22 +2630,38 @@ namespace ts {
             return links.target;
         }
 
-        function markSymbolOfAliasDeclarationIfResolvesToTypeOnly(aliasDeclaration: TypeOnlyCompatibleAliasDeclaration | undefined, resolvesToSymbol: Symbol | undefined) {
-            if (!aliasDeclaration || !resolvesToSymbol) return;
+        /**
+         * If an alias symbol is not itself marked type-only, but resolves to a type-only alias
+         * somewhere in its resolution chain, save a reference to the type-only alias declaration
+         * so the alias _not_ marked type-only can be identified as _transitively_ type-only.
+         * @param aliasDeclaration The alias declaration not marked as type-only
+         * @param resolvesToSymbol A symbol somewhere in the resolution chain of the alias symbol
+         * @param overwriteEmpty Checks `resolvesToSymbol` for type-only declarations even if `aliasDeclaration`
+         * has already been marked as not resolving to a type-only alias. Used when recursively resolving qualified
+         * names of import aliases, e.g. `import C = a.b.C`. If namespace `a` is not found to be type-only, the
+         * import declaration will initially be marked as not resolving to a type-only symbol. But, namespace `b`
+         * must still be checked for a type-only marker, overwriting the previous negative result if found.
+         */
+        function markSymbolOfAliasDeclarationIfResolvesToTypeOnly(aliasDeclaration: Declaration | undefined, resolvesToSymbol: Symbol | undefined, overwriteEmpty?: boolean): boolean {
+            if (!aliasDeclaration || !resolvesToSymbol) return false;
             const sourceSymbol = getSymbolOfNode(aliasDeclaration);
             const links = getSymbolLinks(sourceSymbol);
-            if (links.typeOnlyDeclaration === undefined) {
-                const typeOnly = find(resolvesToSymbol.declarations, isTypeOnlyImportOrExportDeclaration);
+            if (links.typeOnlyDeclaration === undefined || overwriteEmpty && links.typeOnlyDeclaration === false) {
+                resolvesToSymbol = resolvesToSymbol.exports?.get(InternalSymbolName.ExportEquals) ?? resolvesToSymbol;
+                const typeOnly = resolvesToSymbol.declarations && find(resolvesToSymbol.declarations, isTypeOnlyImportOrExportDeclaration);
                 links.typeOnlyDeclaration = typeOnly ?? getSymbolLinks(resolvesToSymbol).typeOnlyDeclaration ?? false;
             }
+            return !!links.typeOnlyDeclaration;
         }
 
-        function markSymbolOfAliasDeclarationIfTypeOnly(aliasDeclaration: TypeOnlyCompatibleAliasDeclaration) {
+        function markSymbolOfAliasDeclarationIfTypeOnly(aliasDeclaration: TypeOnlyCompatibleAliasDeclaration): boolean {
             if (isTypeOnlyImportOrExportDeclaration(aliasDeclaration)) {
                 const symbol = getSymbolOfNode(aliasDeclaration);
                 const links = getSymbolLinks(symbol);
                 links.typeOnlyDeclaration = aliasDeclaration;

```

Comment:
it seems like this has to be called on a declaration before markSymbolOfAliasDeclarationIfResolvesToTypeOnly for the latter to work. Is that correct?

Reply:
Not quite; rather, if this function returns true (indicating that it did something), it’s unnecessary to call the other.

The overall goal is to be able to answer (at the end of `resolveNameHelper`): Supposing an alias is used in a position that will emit to JS, is that alias marked as type-only, or does it refer to something type-only? (And if either is true, we want to get the declaration where the `type` keyword occurred so we can put it in a related span on the error.)

At first, in #36092, I simply answered this question during `resolveNameHelper` by walking the whole import/export chain until I reached either a type-only declaration or the end of the chain (and caching the result). But, to improve performance, I ended up moving that to happen during `resolveAlias`, since that _already_ walks the whole chain. So, you see this function called at every alias declaration that could be marked as type-only. If it _is_ marked as type-only, it doesn’t matter if it _resolves_ to something else type-only or not, because we already know it’s type-only.